### PR TITLE
DOC: added examples for index assignment using Series

### LINKED ISSUE #39845: DOC: Reindexing behaviour of dataframe column-assignment missing
#### Location of the documentation

`pandas.core.indexing.IndexingMixin.loc`
`pandas.DataFrame.__setitem__`

#### Documentation problem

When assigning a `Series` through `df[...] = ...` or `df.loc[...] = `, the `Series`' index is expanded to conform to the `DataFrame`'s, and then values are added according to the index:

```python3
In [1]: df
Out[1]: 
   a
0  1
1  2
2  3

In [2]: df['b'] = pd.Series({1: 'b'})
In [3]: df
Out[3]: 
   a    b
0  1  NaN
1  2    b
2  3  NaN

In [4]: se = pd.Series({2: 'zero', 1: 'one', 0: 'two'})      
# NOTE: Order is preserved and reflected in the Series' order
In [5]: se
Out[5]: 
2    zero
1     one
0     two
dtype: object
In [6]: df['d'] = se
# NOTE: values have been reordered according to the df's index
In [7]: df
Out[8]: 
   a    b    d
0  1  NaN  two
1  2    b  one
2  3  NaN zero
```
(But in contrast:
```python3
In [4]: df['c'] = {1: 'c'} 
<traceback omitted>
ValueError: Length of values (1) does not match length of index (3)
```
)

As far as I can tell, this is not really documented. In the case of `__setitem__` there is *no* API documentation at all, and one is left only with the "Selecting and Indexing Data" guide's examples. In the case of `.loc` there is mention that if using a `Series` as *input*, "The index of the key will be aligned before masking," but this is not what we're doing here. Neither set of examples indicates the behaviour when adding a new column or part of column: the only hints I could find in the guide about [setting with enlargement](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#setting-with-enlargement) added a series whose index was the same as the existing index. This means that it is not clear what order the data will end up in the dataframe and where `NaN`s will be added.

In the case of `.loc` in general the API documentation, although it does exist, is fairly scant. There is a link to the user guide, but personally I think this is pretty important behaviour to document in the reference.

#### Suggested fix for documentation

* Add documentation for `__setitem__` and in particular the behaviour of reindexing `Series`.
* Update documentation for `.loc` to more completely describe the behaviour obtained when *assigning* to `.loc[]`, and include at least one example of assigning to a partial column. Alternatively add this to the user guide. Perhaps something like the following, plus an example like those above:

When assigning a `Series` to a `DataFrame`, either via `.loc` or via the `[]` operator, values of the `Series` will be added to the dataframe according to their index. Values in the `Series` whose label does not appear in the `DataFrame` will not be added, and labels missing from the `Series`' index will be `NaN`. This also means that the order that data appears in the resulting `DataFrame` could be different from the order in the `Series`.

### PR DESCRIPTION
**Pull request by marco-georgaklis**
- [x] closes #39845
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -449,6 +449,37 @@ def loc(self) -> _LocIndexer:
         8          4       5
         9          7       8
 
+        **Set values with a Series**
+
+        Set column values using a Series
```
Comment (mroeschke) [inline pandas/core/indexing.py:454]: Maybe good to add to the end here `aligns along the index values of the Series and DataFrame`

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -449,6 +449,37 @@ def loc(self) -> _LocIndexer:
...
+
+        >>> df.loc[:, 'shield'] = pd.Series({7: 8, 8: 10, 9: 13})
+        >>> df
+           max_speed  shield
+        7          1       8
+        8          4       10
+        9          7       13
+
+        Setting values using a Series with inconsistent indexes aligns the
```
Comment (mroeschke) [inline pandas/core/indexing.py:463]: Then this can just state where the index values are not found in the Series index, the values will be `NaN`

**Review by mroeschke — APPROVED** [review]
LGTM. The failures are unrelated.

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -449,6 +449,38 @@ def loc(self) -> _LocIndexer:
         8          4       5
         9          7       8
 
+        **Set values with a Series**
```
Comment (jreback) [inline pandas/core/indexing.py:452]: this is ok but really not very idiomatic e.g. these should be on the ```__setitem__``` doc-string e.g.
```df['shield'] = ```

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master

File: pandas/core/indexing.py (hunk 3 above)
Comment (datapythonista) [reply pandas/core/indexing.py:452]: Looks like this is the docstring of `.loc`, that's why is using this idiom.

Comment (jreback) [reply pandas/core/indexing.py:452]: sure but these are not idiomatic patterns for *setting a column*; sure if you have a *mask* then its the idimoatic way, but NOT for a column.

**Comment by datapythonista** [comment]
@marco-georgaklis do you mind using a mask when columns are being set in this docstring so we don't show code that we wouldn't use in practice? Would be good if you can do it for the whole dosctring, not only your changes. Thanks!

**Comment by marco-georgaklis** [comment]
I'm not quite sure I understand. Should I be using a mask instead of .loc when the docstring is under the .loc property?

**Comment by datapythonista** [comment]
> I'm not quite sure I understand. Should I be using a mask instead of .loc when the docstring is under the .loc property?

Sorry I wasn't clear. The problem with your changes, is that in pandas we don't use:
```python
>>> df.loc[:, 'shield'] = pd.Series({7: 8, 8: 10, 9: 13})
```

Instead, we would use:
```python
>>> df['shield'] = pd.Series({7: 8, 8: 10, 9: 13})
```

The only case where we'd use `loc` for setting a column, is if there is a mask (a condition on which rows should be updated. Something like:
```python
>>> outstanding_employees = df['performance'] > 80
>>> df.loc[outstanding_employees, 'salary'] = ...
```

So, if you can have a look at the docstring you're updating, check if it makes sense to have an example with a mask, and update it, so the `.loc[:, 'shield']` idiom is not being used, so we don't show code in the docs we wouldn't use, that would be great.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to be stale. If interested in continuing, please merge in the main branch, address the review, and we can reopen.

**Pull request closed without merging**